IPIP/0373-double-hash-dht.md

```diff
@@ -0,0 +1,404 @@
+11. Each DHT server confirms to Content Provider that the Provider Record has been successfully added.
+12. The process is over once Content Provider has received 20 confirmations.
+
+```mermaid
+sequenceDiagram
+    participant CP as Content Provider
+    participant DHT
+    participant Server as DHT Server
+
+    Note left of CP: HASH2 = SHA256(bytes("CR_DOUBLEHASH") || MH)
+
+    CP->>DHT: FIND_PEERS(HASH2)
+    DHT->>CP: [PeerID0, PeerID1, ... PeerID19]
+
+    Note left of CP: EncPeerID = 0x8040 || Nonce || payload_len || AESGCM(MH, Nonce, CPPeerID)
+    Note left of CP: Signature = Sign(privkey, EncPeerID || TS)
+    Note left of CP: ServerKey = SHA256(bytes("CR_SERVERKEY") || MH)
+
+    par Content Provider to the 20 closest DHT Servers to HASH2
+        CP->>Server: HASH2 || EncPeerID || TS || Signature || ServerKey
+    
+    Note right of Server: Verify(pubkey, Signature, EncPeerID) &&<br>TS - time.now() < 48h
+    Note right of Server: On success, add to Provider Store:<br>HASH2 -> ServerKey -> CPPeerID -> [EncPeerID, TS, Signature]
+
+    Server->>CP: Success / Error
+    end
+
+    Note left of CP: Wait for 20 Successes
+```
+
+**Lookup Process**
+1. Client computes `HASH2 = SHA256(bytes("CR_DOUBLEHASH") || MH)` (`MH` is the MultiHash included in the CID).
+2. Client selects a prefix of `HASH2`, `KeyPrefix = HASH2[:l]` for a defined `l` (see [`l` selection](#prefix-length-selection)).
+2. Client finds the closest `PeerID`s to `HASH2` in XOR distance in its Routing Table.
+3. Client sends a DHT lookup request for `KeyPrefix` to these DHT servers. The request contains a flag to specify whether Client wants the `multiaddrs` associated with the `CPPeerID` or not. <!-- TODO: add multiaddrs request flag -->
+4. The DHT servers find the 20 closest `PeerID`s to `KeyPrefix` in XOR distance (see [algorithm](#closest-keys-to-a-key-prefix)). Add these `PeerID`s and their associated multiaddresses (if applicable) to the `message` that will be returned to Client.
+5. The DHT servers search if there are entries matching `KeyPrefix` in their Provider Store.
+6. For all entries `HASH2` of the Provider Store where `HASH2[:len(KeyPrefix)]==KeyPrefix`, add to `message` the following encrypted payload: `EncPeerID || 0x8040 || SERVERNONCE || payload_len || AESGCM(ServerKey, SERVERNONCE, TS || Signature || multiaddrs)`, `SERVERNONCE` being a randomly generated 12-byte array, for `multiaddrs` being the multiaddresses associated with `CPPeerID` (if applicable) if the `multiaddrs` were requested by Client. The `multiaddrs` are taken from the DHT Server's lib2p2 peerstore, and may be stale. If more than `MatchLimit` distinct `HASH2`s match the requested `KeyPrefix`, the DHT Server doesn't return any Provider Record, and adds the number of `HASH2` matching `KeyPrefix` along with its own `MatchLimit` to `message`.
```

Comment:
Questions:
1. We know that `TS` is 32-bits. In a case where addresses are present, how would the client infer where the `Signature` ends and `multiaddrs` start?
2. Pointed out in other comment: would it make sense to use multicodec code prefix for the inner encrypted content? That should then clarify how to parse the unencryted values. 


```suggestion
6. For all entries `HASH2` of the Provider Store where `HASH2[:len(KeyPrefix)]==KeyPrefix`, add to `message` the following encrypted payload: `EncPeerID || 0x2000 || payload_len || SERVERNONCE || AESGCM(ServerKey, SERVERNONCE, TS || Signature || multiaddrs)`, `SERVERNONCE` being a randomly generated 12-byte array, for `multiaddrs` being the multiaddresses associated with `CPPeerID` (if applicable) if the `multiaddrs` were requested by Client. The `multiaddrs` are taken from the DHT Server's lib2p2 peerstore, and may be stale. If more than `MatchLimit` distinct `HASH2`s match the requested `KeyPrefix`, the DHT Server doesn't return any Provider Record, and adds the number of `HASH2` matching `KeyPrefix` along with its own `MatchLimit` to `message`.
```

Reply:
1. The Signature length depends on the [PeerID key type](https://github.com/libp2p/specs/blob/master/peer-ids/peer-ids.md#key-types). Each key type has a fixed length signature output. 
2. What do you refer to as the "inner encrypted content"?

Reply:
1. great thank you. Maybe a quick sentence on this either here or in the terminology section would be useful.
2. `TS || Signature || multiaddrs`. Prefixed with the codc to specify what it is, it then becomes something like: `0x12345 || TS || Signature || multiaddrs` where `0x12345` is the multicodec of DHT routing record.

Reply:
For 2. how about adding this multicodec as below: 

`EncPeerID || MULTICODEC || 0x8040 || SERVERNONCE || payload_len || AESGCM(ServerKey, SERVERNONCE, TS || Signature || multiaddrs)`

This allows to indicate the format of all meta data, unencrypted and encrypted.